### PR TITLE
Jax/Flax implementation of Transformer NMT on WMT'17 en<->de dataset

### PR DESCRIPTION
**Pull request by ckt624**
This model uses Adam optimizer and customized learning rate schedule.

Profiling on 8 V100 GPU instance:  
Total run time: 17652.04 s for 23634 steps, or 1.34 step/second. (Reference: the example WMT in Flax library is 1.35 step/second with the same evaluation frequency).
accuracy: 0.6816954
loss: 1.5500027
bleu: 25.03 (Reference: the example WMT in Flax library is 25.04 after the same steps).
Final wmt_jax score (net training time) : 7119.04 s

### CODE REVIEW

**Comment by georgedahl** [comment]
> This model uses Adam optimizer and customized learning rate schedule.
> 
> Profiling on 8 V100 GPU instance:
> Total run time: 16855.37 s for 24369 steps, or 1.45 step/second. (Reference: the example WMT in Flax library is 1.88 step/second but we have different evaluation frequency).
> accuracy: 0.6855389
> loss: 1.527049
> bleu: 25.233080602490205
> Final wmt_jax score (net training time) : 8386.889772 s

Can we also get a measurement with the same eval frequency? Either setting ours to what Flax uses or modifying the Flax example to use ours?

**Comment by ckt624** [comment]
> > This model uses Adam optimizer and customized learning rate schedule.
> > Profiling on 8 V100 GPU instance:
> > Total run time: 16855.37 s for 24369 steps, or 1.45 step/second. (Reference: the example WMT in Flax library is 1.88 step/second but we have different evaluation frequency).
> > accuracy: 0.6855389
> > loss: 1.527049
> > bleu: 25.233080602490205
> > Final wmt_jax score (net training time) : 8386.889772 s
> 
> Can we also get a measurement with the same eval frequency? Either setting ours to what Flax uses or modifying the Flax example to use ours?

Fixed. Now the speed is very similar.

File: workloads/wmt/wmt_jax/train.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+        ret *= jnp.maximum(0.0,
+                           0.5 * (1.0 + jnp.cos(jnp.pi * (progress % 1.0))))
+      else:
+        raise ValueError("Unknown factor %s." % name)
+    return jnp.asarray(ret, dtype=jnp.float32)
+
+  return step_fn
+
+
+def compute_weighted_cross_entropy(logits,
```
Comment (znado) [inline workloads/wmt/wmt_jax/train.py:81]: we should remove this function, as it seems like a duplicate of `loss_fn` in `workload.py`.

File: workloads/wmt/wmt_jax/train.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+  (_, logits), grad = grad_fn(optimizer.target)
+  grad = jax.lax.pmean(grad, "batch")
+  new_optimizer = optimizer.apply_gradient(grad, learning_rate=lr)
+  metrics = compute_metrics(logits, targets, weights)
+  metrics["learning_rate"] = lr
+
+  return new_optimizer, metrics
+
+
+def eval_step(params, batch, config, label_smoothing=0.0):
```
Comment (znado) [inline workloads/wmt/wmt_jax/train.py:213]: lets remove `label_smoothing` from here and `compute_metrics`, because a non-default value is never actually passed in

File: workloads/wmt/wmt_jax/workload.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    return logits_batch, None
+
+  def loss_fn(
+      self,
+      label_batch: spec.Tensor,  # Dense (not one-hot) labels.
+      logits_batch: spec.Tensor) -> spec.Tensor:
+
+    weights = jnp.where(label_batch > 0, 1.0, 0.0)
+    vocab_size = logits_batch.shape[-1]
+    confidence = 1.0 - self.config.label_smoothing
```
Comment (znado) [inline workloads/wmt/wmt_jax/workload.py:212]: I think we should refactor this to be applied inside `submission.py`, because this is a tunable regularization hyperparameter that the submitter will be providing, so ideally we won't have any submission-specific logic inside `workload.py`. I think this should be pretty easy to do by just applying the label smoothing in `def loss_fn(params): ...` inside `train_step`.

File: workloads/wmt/wmt_jax/workload.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    num_layers=6,
+    qkv_dim=1024,
+    emb_dim=1024,
+    mlp_dim=4096,
+    num_heads=16,
+    dropout_rate=0.1,
+    attention_dropout_rate=0.1,
+    use_bfloat16=True,
+    workdir="./wmt_256",
+    per_device_batch_size=16,
```
Comment (znado) [inline workloads/wmt/wmt_jax/workload.py:46]: we should confirm that this batch_size is not used anywhere, and that only the batch size provided by `submission.py` is used.

File: workloads/wmt/wmt_jax/workload.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    share_embeddings=True,
+    logits_via_embedding=True,
+    num_layers=6,
+    qkv_dim=1024,
+    emb_dim=1024,
+    mlp_dim=4096,
+    num_heads=16,
+    dropout_rate=0.1,
+    attention_dropout_rate=0.1,
+    use_bfloat16=True,
```
Comment (znado) [inline workloads/wmt/wmt_jax/workload.py:44]: this should be `False` for all our workload

File: workloads/wmt/wmt_jax/workload.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    dataset_name="wmt17_translate/de-en",
+    eval_split="test",
+    reverse_translation=True,
+    beam_size=4,
+    num_eval_steps=20,
+    num_predict_steps=-1,
+    learning_rate=0.0625,
+    warmup_steps=1000,
+    label_smoothing=0.1,
+    weight_decay=0.0,
```
Comment (znado) [inline workloads/wmt/wmt_jax/workload.py:31]: `learning_rate`, `label_smoothing`, and `weight_decay` should all be removed from the configs, because these will be submission provided tunable hpraams

**Review by znado — COMMENTED** [review]
Overall this is looking good, I need to finish going through some of the files, but I think that can wait until this set of discussions is resolved!

File: workloads/wmt/wmt_jax/train.py (hunk 1 above)
Comment (ckt624) [reply workloads/wmt/wmt_jax/train.py]: I reused it in loss_fn

File: workloads/wmt/wmt_jax/train.py [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+
+# Primary training / eval / decode step functions.
+# -----------------------------------------------------------------------------
+
+
+def train_step(optimizer,
+               batch,
+               config,
+               learning_rate_fn,
+               label_smoothing=0.0,
```
[the remark this answers is not in the document]
Comment (ckt624) [reply workloads/wmt/wmt_jax/train.py]: Yes, label_smoothing is from config and dropout_rng is from update_params.

File: workloads/wmt/wmt_jax/submission.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+  # Replicate optimizer.
+  optimizer = jax_utils.replicate(optimizer)
+
+  learning_rate_fn = train.create_learning_rate_scheduler(
+      base_learning_rate=hyperparameters.learning_rate,
+      warmup_steps=workload.config.warmup_steps)
+
+  # compile multidevice versions of train/eval/predict step and cache init fn.
+  p_train_step = jax.pmap(
```
[the remark this answers is not in the document]
Comment (ckt624) [reply workloads/wmt/wmt_jax/submission.py:49]: It seems that then we need to explicitly deal with replica of optimizers? The existing method doesn't require replica/unreplica optimizers manually.

File: workloads/wmt/wmt_jax/train.py [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+
+
+# Primary training / eval / decode step functions.
+# -----------------------------------------------------------------------------
+
+
+def train_step(optimizer,
+               batch,
+               config,
+               learning_rate_fn,
```
[the remark this answers is not in the document]
Comment (ckt624) [reply workloads/wmt/wmt_jax/train.py]: Same above.

File: workloads/wmt/wmt_jax/submission.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  Return a tuple of input label batches.
+  """
+  del optimizer_state
+  del current_param_container
+  del global_step
+  del rng
+  del hyperparameters
+  del workload
+
+  return common_utils.shard(jax.tree_map(np.asarray, next(input_queue))), None
```
[the remark this answers is not in the document]
Comment (ckt624) [reply workloads/wmt/wmt_jax/submission.py:234]: Yes.

File: workloads/wmt/wmt_jax/workload.py [hunk 11]
```diff
@@ -0,0 +1,593 @@
...
+    return selected_raw_input_batch, selected_label_batch
+
+  def preprocess_for_eval(self, raw_input_batch: spec.Tensor,
+                          train_mean: spec.Tensor,
+                          train_stddev: spec.Tensor) -> spec.Tensor:
+    del train_mean
+    del train_stddev
+    return raw_input_batch
+
+  def init_model_fn(self, rng: spec.RandomState) -> spec.ModelInitState:
```
[the remark this answers is not in the document]
Comment (ckt624) [reply workloads/wmt/wmt_jax/workload.py:361]: This function is required by spec. And the return value is used in submission_runner.

File: workloads/wmt/wmt_jax/workload.py [hunk 12]
```diff
@@ -0,0 +1,593 @@
...
+
+  # Utils for prediction and BLEU calculation
+  # -----------------------------------------------------------------------------
+
+  def pad_examples(self, x, desired_batch_size):
+    """Expand batch to desired size by repeating last slice."""
+    batch_pad = desired_batch_size - x.shape[0]
+    return np.concatenate([x, np.tile(x[-1], (batch_pad, 1))], axis=0)
+
+  def per_host_sum_pmap(self, in_tree):
```
[the remark this answers is not in the document]
Comment (ckt624) [reply workloads/wmt/wmt_jax/workload.py:199]: bleu_matches is a tuple of 4 1d arrays. It's not obvious to me how the function can be simplified.

File: workloads/wmt/wmt_jax/submission.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ckt624) [reply workloads/wmt/wmt_jax/submission.py:165]: It seems most natural to create p_train_step in initialization rather than update_param so that we don't assign learning_rate_scheduler and config to train_step every time.

**Review by znado — APPROVED** [review]
